source/client-side-encryption/tests/README.rst

```diff
@@ -2188,3 +2188,142 @@ attached to the virtual machine.
+the `automatic Azure credentials section <auto-azure_>`_ with the following
+changes:
+
+1. Instead of the standard IMDS TCP endpoint of `169.254.169.254:80`,
+   communicate with the running ``fake_azure`` HTTP server.
+
+2. For each test case, the behavior of the server may be controlled by attaching
+   an additional HTTP header to the sent request: ``X-MongoDB-HTTP-TestParams``.
+
+
+Case 1: Success
+```````````````
+
+Do not set an ``X-MongoDB-HTTP-TestParams`` header.
+
+Upon receiving a response from ``fake_azure``, the driver must decode the
+following information:
+
+1. HTTP status will be ``200 Okay``.
+2. The HTTP body will be a valid JSON string.
+3. The access token will be the string ``"magic-cookie"``.
+4. The expiry duration of the token will be seventy seconds.
+5. The token will have a resource of ``"https://vault.azure.net"``
+
+
+Case 2: Empty JSON
+``````````````````
+
+This case addresses a server returning valid JSON with invalid content.
+
+Set ``X-MongoDB-HTTP-TestParams`` to ``case=empty-json``.
+
+Upon receiving a response:
+
+1. HTTP status will be ``200 Okay``
+2. The HTTP body will be a valid JSON string.
+3. There will be no access token, expiry duration, or resource.
+
+The test case should ensure that this error condition is handled gracefully.
+
+
+Case 3: Bad JSON
+````````````````
+
+This case addresses a server returning malformed JSON.
+
+Set ``X-MongoDB-HTTP-TestParams`` to ``case=bad-json``.
+
+Upon receiving a response:
+
+1. HTTP status will be ``200 Okay``
+2. The response body will contain a malformed JSON string.
+
+The test case should ensure that this error condition is handled gracefully.
+
+
+Case 4: HTTP 404
+````````````````
+
+This case addresses a server returning a "Not Found" response. This is
+documented to occur spuriously within an Azure environment.
+
+Set ``X-MongoDB-HTTP-TestParams`` to ``case=404``.
+
+Upon receiving a response:
+
+1. HTTP status will be ``404 Not Found``.
+2. The response body is unspecified.
+
+The test case should ensure that this error condition is handled gracefully.
+
+
+Case 5: HTTP 500
+````````````````
+
+This case addresses an IMDS server reporting an internal error. This is
+documented to occur spuriously within an Azure environment.
+
+Set ``X-MongoDB-HTTP-TestParams`` to ``case=500``.
+
+Upon receiving a response:
+
+1. HTTP status code will be ``500``.
+2. The response body is unspecified.
+
+The test case should ensure that this error condition is handled gracefully.
+
+
+Case 6: Slow Response
+`````````````````````
+
+This case addresses an IMDS server responding very slowly. Drivers should not
+halt the application waiting on a peer to communicate.
+
+Set ``X-MongoDB-HTTP-TestParams`` to ``case=slow``.
+
+The HTTP response from the ``fake_azure`` server will take at least 1000 seconds
+to complete. The client should report a timeout within a few seconds.
+
+
+Case 7: Large Response
```

Comment:
should we have a case that tests caching?

Reply:
Additional cases are to be defined in a subsequent changeset, and I think cache handling may be better checked in the true environment.